File: python/merge_sort.py
my_list = [3, 1, 11, 9, 5, 4, 2, 7, 21]


def merge_sort(num_list):
    if len(num_list) <= 1:
        return num_list
    mid = len(num_list) // 2
    left = merge_sort(num_list[:mid])
    right = merge_sort(num_list[mid:])

    sorted_list = []
    left_idx = right_idx = 0
    while left_idx < len(left) and right_idx < len(right):
        if left[left_idx] < right[right_idx]:
            sorted_list.append(left[left_idx])
            left_idx += 1
        else:
            sorted_list.append(right[right_idx])
            right_idx += 1
    sorted_list += left[left_idx:]
    sorted_list += right[right_idx:]
    return sorted_list


answer = merge_sort(my_list)
print(answer)

File: javascript/mergeSort.js
const numList = [3, 6, 5, 9, 1, 4, 2, 8, 7];

function mergeSort(nums) {
  if (nums.length < 2) {
    return nums;
  }
  const mid = Math.floor(nums.length / 2);
  const left = nums.slice(0, mid);
  const right = nums.slice(mid);
  return merge(mergeSort(left), mergeSort(right));

  function merge(left, right) {
    const resultArray = [];
    let leftIdx = 0;
    let rightIdx = 0;
    while (leftIdx < left.length && rightIdx < right.length) {
      if (left[leftIdx] < right[rightIdx]) {
        resultArray.push(left[leftIdx]);
        leftIdx++;
      } else {
        resultArray.push(right[rightIdx]);
        rightIdx++;
      }
    }
    return resultArray.concat(left.slice(leftIdx), right.slice(rightIdx));
  }
}
console.log(mergeSort(numList));
